### PR TITLE
HPCC-17294 Improve CCsvPartitioner::getRecordStructure() assertex message.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add more error checking and proper messages

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Tested manually with good and bad terminator parameter and small (<headerSize) and large files.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: dali/ft/daftformat.cpp [hunk 1]
```diff
@@ -702,19 +702,33 @@ void CCsvPartitioner::getRecordStructure(StringBuffer & _recordStructure)
...
         const byte *buffer = bufferBase();
-        ensureBuffered(headerSize);
-        assertex((headerSize == 0) || (numInBuffer != bufferOffset));
+        unsigned processSize = bufferSize;
+
+        if (!ensureBuffered(headerSize) && (numInBuffer < headerSize))
+        {
+            // This means the stream/file is shortest than the headerSize, however it can hold
+            // valid records
+            LOG(MCdebugProgressDetail, unknownJob, "It is possible the file is shorter (%d bytes) than the header size(%d bytes)", numInBuffer, headerSize);
```
Comment (jakesmith) [inline dali/ft/daftformat.cpp]: Makes sense I think, file shorter than maxrecord size basically..
I don't think logging this is appropriate though, a comment would suffice here.

File: dali/ft/daftformat.cpp [hunk 2]
```diff
@@ -702,19 +702,33 @@ void CCsvPartitioner::getRecordStructure(StringBuffer & _recordStructure)
...
         {
-            LOG(MCdebugProgressDetail, unknownJob, "First record size %d (0x%08x) is larger than the buffer size: %d", size, size, bufferSize);
+            LOG(MCdebugProgressDetail, unknownJob, "First record size %d (0x%08x) is larger than the buffer size: %d. Please check the separator and terminator parameters.", size, size, bufferSize);
             throwError2(DFTERR_WrongSplitRecordSize, size, size);
         }
+        else if (isFirstRow)
+        {
+            // Possibly reached the end of the file but the processing is still in first row
+            // so no terminator detected
+            LOG(MCdebugProgressDetail, unknownJob, "%d (0x%08x) bytes is processed, but there is no terminator found. Please check the separator and terminator parameters.", size, size);
```
Comment (jakesmith) [inline dali/ft/daftformat.cpp]: Should be "bytes processed"

**Comment by jakesmith** [comment]
@AttilaVamos - looks good, only minor comments.

File: dali/ft/daftformat.cpp [hunk 3]
```diff
@@ -702,19 +702,33 @@ void CCsvPartitioner::getRecordStructure(StringBuffer & _recordStructure)
     if (nullptr == strstr(recordStructure.str(),"END;"))
     {
         const byte *buffer = bufferBase();
-        ensureBuffered(headerSize);
-        assertex((headerSize == 0) || (numInBuffer != bufferOffset));
+        unsigned processSize = bufferSize;
+
+        if (!ensureBuffered(headerSize) && (numInBuffer < headerSize))
+        {
+            // This means the stream/file is shortest than the headerSize, however it can hold
```
Comment (AttilaVamos) [inline dali/ft/daftformat.cpp]: Updated

File: dali/ft/daftformat.cpp (hunk 2 above)
Comment (AttilaVamos) [inline dali/ft/daftformat.cpp]: Updated

File: dali/ft/daftformat.cpp (hunk 1 above)
Comment (AttilaVamos) [inline dali/ft/daftformat.cpp]: Removed

**Comment by AttilaVamos** [comment]
@jakesmith updated, please check.

**Comment by jakesmith** [comment]
@richardkchapman - good to merge.

**Pull request merged**